Chapter 4 Task 18

tasiq opened the pull request.

tasiq committed:
Chapter 4 Task 18

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,46 @@
+
+    public static int[] doCocktailSort(int[] array) {
+
+        int length = array.length;
```
эта переменная не нужна.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,46 @@
+            if (i != 0 && countMovesLeft == 0 && countMovesRight == 0) {
+                break;
+            }
+            for (int k = length - 1 - i; k > 0; k--) {
```
неверное условие про k > 0 с каждым прохождения минимального элемента влево цикл должен становиться короче.

EvgeneZhurov replied:
итератор 2 уровня вложенности в цикле называют j

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,46 @@
+
+        int length = array.length;
+
+        for (int i = 0; i < length; i++) {
```
i < length вторую половину лины цикл не выполнит никогда.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,46 @@
+        for (int i = 0; i < length; i++) {
+            int countMovesRight = 0;
+            int countMovesLeft = 0;
+            for (int j = i; j < length - 1; j++) {
```
length - 1   точно так? с каждым прохождением максимального элемента вправо цикл должен становиться короче.

tasiq committed:
Chapter 4 Task 18:

- убрана лишняя переменная arrayLength, вместо нее обращение к длине массива напрямую;
- общее количество проходов по массиву туда и обратно исправлено с array.length на array.length/2, округленное вверх;
- циклы второго уровня вложенности более не захватывают крайние значения массива, отсортированные на предыдущем шаге;
- в цикле второго уровня вложенности k переименовано на j;
- исправлено форматирование.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,51 @@
+                }
+            }
+
+            if (i != 0 && countMovesLeft == 0 && countMovesRight == 0) {
```
countMovesLeft == 0  - это условие никогда не выполнится. Либо цикл проедет мимо на 1 итерации цикла и выйдет из него в нижнем if, либо countMovesLeft != 0. По этому идея подсвечивает, что условие бесполезно. Обращайте внимание на варнинги.

tasiq replied:
Ох, это было очень глупо. Да, я обращаю, но в этот раз почему-то решил, что глупая IDEA меня не поняла. И даже ковырялся дебагом, но сейчас понимаю, что проверял сценарии, в которых как раз ожидался countMovesLeft  == 0, а надо было проверить и countMovesLeft  != 0 тоже.
Очень наглядно было, больше так не ошибусь, спасибо.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,51 @@
+            int countMovesLeft = 0;
+
+            for (int j = i; j < array.length - 1 - i; j++) {
+
```
сразу после { можно не пропускать строку.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,51 @@
+            int countMovesRight = 0;
+            int countMovesLeft = 0;
+
+            for (int j = i; j < array.length - 1 - i; j++) {
```
когда максимальный элемент представился на самый правый индекс, перед тем как сравнивать минимальные элементы нужно отступить влево.

tasiq committed:
Chapter 4 Task 18:

- при проходе массива в обратную сторону перебор начинается не с конца массива, а с последнего из неотсортированных элементов;
- исправлена ошибка в счетчиках перестановок, цикл завершится на проходе n (кроме самого первого), по итогам которого не будет ни одной перестановки.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -5,43 +5,45 @@
 
+
         for (int i = 0; i < Math.ceil((double) array.length / 2); i++) {
 
```
сразу после { и перед } обычно не ставят пустые строки, а 2 пустые стоки считается моветоном.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -5,43 +5,45 @@
 
-            int countMovesRight = 0;
-            int countMovesLeft = 0;
+            int countSwap = 0;
```
это на самом деле завуалированная логическая переменная, которая отвечает на вопрос, отсортирован ли массив. Лучше и делать ее логической без скрытого смысла.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -5,43 +5,45 @@
             }
 
-            if (countMovesLeft == 0 && countMovesRight == 0) {
+            if (countSwap == 0) {
```
это условие можно попробовать вставить в логическое условие выхода в самом for ( ; ... && countSwap == 0 ; )

tasiq replied:
Магия. Но понял идею, кажется.

Зато забыл, почему решил, что надо продолжать сортировку, если после самого первого прохода не было ни одной перестановки(

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -5,43 +5,45 @@
+            countSwap = 0;
+
+            for (int j = array.length - 2 - i; j > i; j--) {
 
```
здесь не нужна пустая строка

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -5,43 +5,45 @@
         System.out.println(Arrays.toString(doCocktailSort(array)));
     }
 
     public static int[] doCocktailSort(int[] array) {
```
избегайте слова do, и так понятно, что метод нужен для действия. Когда речь заходит о каких то известных названиях, можно обойтись и без глагола.

EvgeneZhurov replied:
Сортировка вводимого массива будет произведена и без возврата самого массива, можно заменить на void.

tasiq committed:
Chapter 4 Task 18:

- изменен выход из сортировки, условие выхода добавлено в условия цикла верхнего уровня и во второй цикл второго уровня.
- первый цикл второго уровня не менялся, дополнительная проверка не нужна. Если бы isSorted был true, то до этого цикла не дойдет, потому что цикл верхнего уровня прекратился бы и не пустил дальше.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -1,7 +1,46 @@
+        boolean isSorted = false;
+
+        for (int i = 0; i < Math.ceil((double) array.length / 2) && !isSorted; i++) {
+            int countMoves = 0;
```
эта переменная лишняя, лучше ее выбросить из кода и поставить isSorted = true;  а внутри if присваивать логической переменной false. Суть проста, если код внутри if не исполняется, значит массив отсортирован. Сейчас переменная countMoves выглядит ненужным костылем

tasiq committed:
Chapter 4 Task 18:
- условие досрочного выхода из перебора значений, если массив уже отсортирован, удалено из условий цикла for. Цикл прерывается через break, если за проход не было ни одной перестановки.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task18/Main.java:
```diff
@@ -5,41 +5,40 @@
             }
-            if (countMoves == 0) {
-                isSorted = true;
+            if (isSorted) {
```
лучше отделять блоки кода пустыми строками с обеих сторон
